README.md

```diff
@@ -0,0 +1,400 @@
+# Unsize and CoerceUnsize v2
+
+This repo experiments with a new design for the `Unsize` and `CoerceUnsized` traits, making them more flexible and applicable to more types.
+The motivation is to see how flexible the traits could be made and whether this flexibility pays off or not.
+An idea that kicked this off was to have `Vec<T>: Unsize<[T]>`, which requires user code to run to be able to adjust the pointer for the data.
+Today `Vec<T>` implements `Deref<Target = [T]>` for convenience as this allows dispatching slice methods on objects of `Vec<T>`, ye `Deref` was meant for smart pointers instead. What this impl really feels like is more akin to unsizing. Obviously the `Deref` impl cannot be revoked from `Vec<T>` anymore, but assuming an unsizing impl would've been the proper call this would then also imply that unsizing would have to be introduced into autoderef (instead of just special casing array unsizing).
+
+What follows here is the accompanied pre-RFC:
+
+# RFC
+
+## Unsize- Feature Name: `unsizev2`
+- Start Date: (fill me in with today's date, YYYY-MM-DD)
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Move the unsizing logic out of the compiler into library source, allowing for a more flexible design of the features.
+
+# Motivation
+[motivation]: #motivation
+
+Currently unsizing in Rust is very rigid, only allowed in very specific scenarios permitted by the rules surrounding the current `Unsize` and `CoerceUnsized` traits and it's automatic implementations by the compiler.
+
+This on one hand, has the downside of being very magical, as the majority of the logic happens inside the compiler opposed to the source code and it also prevents certain unsizing implementations from being doable today
+
+This RFC attempts to make these rules more flexible by also allowing user implementations of the traits that define how the metadata is derived and composed back into unsized objects.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
```

Comment:
Maybe a brief overview of the relationship between unsize and coerceunsize would be helpful in the beginning?

Reply:
I guess this was already clear to you Lukas, but I just realized that `Unsize` converts from `T::Metadata` to `U::Metadata` and `CoerceUnsize` converts from `T` to `U`.

```
 ___________                 ___________
|           | CoerceUnsize  |           |
|     T     |-------------->|     U     |
|           |               |           |
| Metadata-------------------->Metadata |
|___________|    Unsize     |___________|
```

Reply:
lol, I just read that this is what you are saying in the reference level explaination

Reply:
Ye, I am having trouble separating the guide vs reference explanations. In part because this RFC kind of talks like introducing these things while also comparing to the status quo. Its a bit odd.